### DISCUSSION TITLE
API service per finalizzare e inviare una fattura

### DISCUSSION THREAD

**Discussion by tmoretto17** [fattureincloud/api#40] [Q&A] [answered]
Apro il seguente thread in quanto non riesco a individuare il servizio da invocare per finalizzare e inviare una fattura corrispettivo all'interazione eseguibile dal FE di fatture in cloud.
Qualcuno può aiutarmi?

**Comment by valmoz** [comment, marked as answer]
Buongiorno,
può guardare qui: https://developers.fattureincloud.it/docs/guides/invoice-creation

Ci faccia sapere se ho risposto alla sua domanda, oppure ci dia ulteriori dettagli.

Grazie
Mauro

**Reply by tmoretto17** [reply]
Buongiorno,
Io ho creato senza problemi la proforma e trasformato la stessa in fattura. Ora dal FE di fatture in cloud con il tasto finalizza e invia questa da quello che ho capita viene inviata al cliente e registrata presso l'agenzia delle entrate. Mi chiedevo come implementare tale logica.
Devo in sequenza innovare i servizi:
    POST /c/{company_id}/issued_documents/{document_id}/e_invoice/send
    POST /c/{company_id}/issued_documents/{document_id}/email
Grazie per la pronta risposta!

**Reply by valmoz** [reply]
Buongiorno @tmoretto17 
credo che ti sia risposto da solo :D in caso contrario, facci sapere :)

I metodi che hai citato sono:
- [Send E-Invoice](https://developers.fattureincloud.it/api-reference/#post-/c/-company_id-/issued_documents/-document_id-/e_invoice/send)
- [Schedule Email](https://developers.fattureincloud.it/api-reference/#post-/c/-company_id-/issued_documents/-document_id-/email)

Se servono ulteriori informazioni, puoi leggere questa guida:
https://developers.fattureincloud.it/docs/guides/e-invoice-management

Grazie
Mauro

**Reply by tmoretto17** [reply]
Approfitto del thread per chiedere se per trasformare una proforma in fattura sia solo necessario invocare il servizio
- PUT /c/{company_id}/issued_documents/{document_id}
passando il campo "type" del body sestato a "invoice" invece che "proforma"

**Reply by emajo** [reply]
Ciao @tmoretto17 attualmente non stiamo esponendo la chiamata per la trasformazione dei documenti, stiamo lavorando per esporla il più presto possibile.

Grazie,
Emanuele

**Reply by tmoretto17** [reply]
Quindi attualmente un workaround potrebbe essere mappare creare una nuova fattura a partire dai metadati di una proforma, in modo da ottenere la numerazione automatica. E inserire la frase di cortesia:
- In riferimento alla seguente proforma:
  Proforma n. 321 del 29/04/2022
  Copia del documento elettronico disponibile nella Sua area riservata dell'Agenzia delle Entrate

C'è qualche controindicazione in una gestione di questo tipo in attesa che venga esposto il servizio di trasformazione di una proforma in fattura?

**Reply by valmoz** [reply]
Ciao @tmoretto17 
al momento la nostra webapp (che utilizza le stesse API) effettua le seguenti chiamate:
- Transform (la chiamata ancora da esporre) per recuperare i dettagli della fattura da visualizzare sulla pagina di trasformazione
- Create Issued Document per creare effettivamente la fattura

N.B.: la transform non effettua quindi la trasformazione vera e propria del documento, ma è una GET che si occupa solo di restituire la precompilazione della pagina. La trasformazione effettiva avviene sempre tramite la Create Issued Document.

L'unica differenza sta nel fatto che l'app web aggiunge alcuni campi particolari quando la chiamata viene effettuata a seguito della trasformazione, in modo tale da scatenare alcune funzionalità che non vengono eseguite con la creazione normale (ad esempio la rimozione della proforma originale, se necessario).
Riassumendo, non dovrebbero esserci grandi controindicazioni nell'applicare il flusso da lei proposto (che è praticamente lo stesso applicato da noi), ma le chiederei comunque di effettuare gli opportuni controlli perché senza i parametri aggiuntivi probabilmente sarà richiesto di effettuare alcune operazioni a mano.

Stiamo comunque pianificando di esporre e documentare il metodo transform, ma non abbiamo ancora delle tempistiche precise.

**Reply by tmoretto17** [reply]
In realtà ho già implementato un FE che permette di aggiungere i campi che sono necessari per poi effettuare la chiamata al servizio:
[Send E-Invoice](https://developers.fattureincloud.it/api-reference/#post-/c/-company_id-/issued_documents/-document_id-/e_invoice/send)
Ergo se il flusso che ho descritto è analogo a quello da voi applicato mi basterà semplicemente introdurre la cancellazione della proforma e in fase di trasformazione dal nostro CRM obbligare l'utente a inserire i dati necessari per la corretta creazione della fattura.
